### PR TITLE
Logging steps

### PR DESCRIPTION
**Pull request by dividedby0**
Logging steps allow to group and nest logs.
Method `startStep` of `LoggerClientAbstract`, reports log with type `step` and creates a new step. It is possible to call `startStep` multiple times, then each next step will have previous as parent. All reports logged after last invocation of `startStep` will have last step as parent and, instead of immediate broadcast to `LoggerServer`, will be added to batch.
Method `endStep` pops last step from stack. When last step popped out of stack, client sends batch to `LoggerServer`

### CODE REVIEW

File: packages/testring-logger/interfaces.d.ts [hunk 1]
```diff
@@ -3,8 +3,9 @@ import { LogTypes } from './src/structs';
 export interface ILogEntry {
     time: Date,
     type: LogTypes,
-    nestingLevel: number,
     logLevel: number,
     content: string[],
-    formattedMessage: string
+    formattedMessage: string,
+    stepUid?: string,
+    parentStep: string | null,
```
Comment (johnthecat) [inline packages/testring-logger/interfaces.d.ts]: please, make `parentStep?: string`

File: packages/testring-logger/package.json [hunk 2]
```diff
@@ -12,7 +12,9 @@
   "dependencies": {
     "@testring/transport": "0.1.0",
     "@testring/types": "0.1.0",
-    "@testring/pluggable-module": "0.1.0"
+    "@testring/pluggable-module": "0.1.0",
+    "@types/nanoid": "^1.0.0",
+    "nanoid": "^1.0.3"
```
Comment (johnthecat) [inline packages/testring-logger/package.json]: use `^1.0.4` plz))))

File: packages/testring-logger/src/abstract-logger-client.ts [hunk 3]
```diff
@@ -4,83 +4,136 @@ import { transport } from '@testring/transport';
...
-        return util.format(
-            `[${time.toLocaleTimeString()}]`, ...content
-        );
+    protected abstract broadcast(messageType: string, payload: any): void;
+
+    protected stepStack: Array<string> = [];
+
+    protected logBatch: Array<ILogEntry> = [];
+
+    protected get currentStep() {
```
Comment (johnthecat) [inline packages/testring-logger/src/abstract-logger-client.ts]: please, make as method

File: packages/testring-logger/src/abstract-logger-client.ts [hunk 4]
```diff
@@ -4,83 +4,136 @@ import { transport } from '@testring/transport';
...
+        return this.stepStack[this.stepStack.length - 1] || null;
     }
 
-    protected abstract broadcast(messageType: string, payload: any): void;
+    protected get previousStep() {
```
Comment (johnthecat) [inline packages/testring-logger/src/abstract-logger-client.ts]: as above

File: packages/testring-logger/src/abstract-logger-client.ts [hunk 5]
```diff
@@ -4,83 +4,136 @@ import { transport } from '@testring/transport';
...
+            this.logBatch.push(logEntry);
+        } else {
+            this.broadcast(
+                LoggerMessageTypes.REPORT,
+                logEntry,
+            );
+        }
+    }
+
+    protected sendLogBatch(): void {
```
Comment (johnthecat) [inline packages/testring-logger/src/abstract-logger-client.ts]: `sendBatchedLog`

File: packages/testring-logger/interfaces.d.ts (hunk 1 above)
Comment (dividedby0) [reply packages/testring-logger/interfaces.d.ts]: We decided that it's semantically right

**Pull request merged**